sdk/go/pulumi/resource_test.go

```diff
@@ -317,3 +348,231 @@ func TestResourceOptionMergingReplaceOnChanges(t *testing.T) {
+			return out.(AnyOutput)
+		}
+
+		checkDeps := func(name string, dependsOn AnyOutput, expectedDeps ...Resource) {
+			res := newTestRes(t, ctx, name, DependsOnOutput(dependsOn))
+			assertHasDeps(t, ctx, depTracker, res, expectedDeps...)
+		}
+
+		dep1 := newTestRes(t, ctx, "dep1")
+		dep2 := newTestRes(t, ctx, "dep2")
+		dep3 := newTestRes(t, ctx, "dep3")
+
+		out := outputDependingOnResource(dep1, true).
+			ApplyT(func(int) Resource { return dep2 }).(ResourceOutput)
+
+		checkDeps("r1", anyOut([]Resource{dep1, dep2}), dep1, dep2)
+		checkDeps("r2", anyOut([]ResourceInput{out}), dep1, dep2)
+		checkDeps("r3", anyOut([]interface{}{out, dep3}), dep1, dep2, dep3)
+
+		dep4 := newTestRes(t, ctx, "dep4")
+		out4 := outputDependingOnResource(dep4, true).
+			ApplyT(func(int) AnyOutput { return anyOut([]Resource{dep1, dep2}) }).(AnyOutput)
+		checkDeps("r4", out4, dep1, dep2, dep4)
+
+		return nil
+	}, WithMocks("project", "stack", &testMonitor{}))
+	assert.NoError(t, err)
+}
+
+func assertHasDeps(
+	t *testing.T,
+	ctx *Context,
+	depTracker *dependenciesTracker,
+	res Resource,
+	expectedDeps ...Resource) {
+
+	name := res.getName()
+	resDeps := depTracker.dependencies(urn(t, ctx, res))
+
+	var expDeps []URN
+	for _, expDepRes := range expectedDeps {
+		expDep := urn(t, ctx, expDepRes)
+		expDeps = append(expDeps, expDep)
+		assert.Containsf(t, resDeps, expDep, "Resource %s does not depend on %s",
+			name, expDep)
+	}
+
+	for _, actualDep := range resDeps {
+		assert.Containsf(t, expDeps, actualDep, "Resource %s unexpectedly depend on %s",
+			name, actualDep)
+	}
+}
+
+func outputDependingOnResource(res Resource, isKnown bool) IntOutput {
+	out := newIntOutput()
+	out.resolve(0, isKnown, false /* secret */, []Resource{res})
+	return out
+}
+
+func newTestRes(t *testing.T, ctx *Context, name string, opts ...ResourceOption) Resource {
+	var res testRes
+	err := ctx.RegisterResource(fmt.Sprintf("test:resource:%stype", name), name, nil, &res, opts...)
+	assert.NoError(t, err)
+	return &res
+}
+
+func urn(t *testing.T, ctx *Context, res Resource) URN {
+	urn, _, _, err := res.URN().awaitURN(ctx.ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return urn
+}
+
+type dependenciesTracker struct {
+	dependsOn *sync.Map
+}
+
+func (dt *dependenciesTracker) dependencies(resource URN) []URN {
+	val, ok := dt.dependsOn.Load(resource)
+	if !ok {
+		return nil
+	}
+	urns, ok := val.([]URN)
+	if !ok {
+		return nil
+	}
+	return urns
+}
+
+func trackDependencies(ctx *Context) *dependenciesTracker {
+	dependsOn := &sync.Map{}
+	m := newInterceptingResourceMonitor(ctx.monitor)
+	m.afterRegisterResource = func(in *pulumirpc.RegisterResourceRequest,
+		resp *pulumirpc.RegisterResourceResponse,
+		err error) {
+		var deps []URN
+		for _, dep := range in.GetDependencies() {
+			deps = append(deps, URN(dep))
+		}
+		dependsOn.Store(URN(resp.Urn), deps)
+	}
+	ctx.monitor = m
+	return &dependenciesTracker{dependsOn}
+}
+
+type interceptingResourceMonitor struct {
```

Comment:
Brute force attack here on the problem of intercepting engine calls to reconstruct dependency map "as reported to the enigne". Please LMK if there are other more elegant ways. Perhaps build this into mockEngine instead of mockMonitor?

Reply:
Ah, is this necessary b/c we don't include dependency info in `MockResourceArgs`?

Reply:
Right, the other way would be to package this into MockResourceArgs, simpler implementation. That'd be a public interface change and up for consideration if we want that public change or not, I did not want to tag that along here. OK to tackle separately or you'd prefer to go ahead with extending MockResourceArgs and I can make this PR depend on it?